nurse_scheduling.py

```diff
@@ -0,0 +1,161 @@
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This code includes an implementation of the algorithm described in Ikeda, 
+# K., # Nakamura, Y. & Humble, T.S. Application of Quantum Annealing to Nurse 
+# Scheduling Problem. Sci Rep 9, 12837 (2019). 
+# https://doi.org/10.1038/s41598-019-49172-3, © The Author(s) 2019, use of 
+# which is licensed under a Creative Commons Attribution 4.0 International 
+# License (To view a copy of this license, visit 
+# http://creativecommons.org/licenses/by/4.0/).
+
+from dwave.system import LeapHybridSampler
+from dimod import BinaryQuadraticModel
+
+# count nurses n = 1 ... N_NURSES
+# count scheduling days as d = 1 ... N_DAYS
+# binary variable q_nd is the assignment of nurse n to day d
+# a is a positive correlation coefficient for implementing the hard nurse
+# constraint - value provided by Ikeda, Nakamura, Humble
+a = 3.5
+N_NURSES = 3
+N_DAYS = 11
+SIZE = N_DAYS * N_NURSES
+
+# Hard shift constraint: at least one nurse is working each day
+# implemented by penalizing any nurses scheduled on two successive
+# days
+
+# Hard nurse constraint: no nurse works two or more consecutive days
+# Lagrange parameter, for hard nurse constraint, on workforce and effort
+# Workforce function W(d) - set to a constant for now
+# Effort function E(n) - set to a constant for now
+lagrange_parameter = 1.3
+W = 1
+E = 1
+
+# Soft nurse constraint: all nurses should have approximately even work
+#                        schedules
+# Lagrange parameter, for shift constraints, on work days
+# preference function G - set to a constant for now
+# Minimum duty days F - the number of work days that each nurse wants
+# to be scheduled. At present, each will do the minimum on average.
+# The parameter gamma's value suggested by Ikeda, Nakamura, Humble
+gamma = 0.3
+G = 1
+F = int(N_DAYS/N_NURSES)
+
+
+def indx(n, d):
+    return n * N_DAYS + d
+
+
+def get_nurse_and_day(indx):
+    return (int(indx / N_DAYS), indx % N_DAYS)
+
+
+# Hard shift constraint - does not have Lagrange parameter - J matrix
+# symmetric, real-valued interaction matrix J, whereas all terms are
+# a or zero.
+# composite indices i(n, d) and j(n, d) as functions of n and d
+# In other words, the indices are not n and d. They are kind of
+# (n+d)**2, and we need to take that into account. This seems to be
+# a clean way to do it.
+# J_i(n,d)j(n,d+1) = a and 0 otherwise.
+J = {}
+for nurse_day_1 in range(SIZE):
+    for nurse_day_2 in range(SIZE):
+        J[nurse_day_1, nurse_day_2] = 0
+        if int(nurse_day_1 / N_DAYS) == int(nurse_day_2 / N_DAYS) and nurse_day_2 == nurse_day_1 + 1:
+            J[nurse_day_1, nurse_day_2] = a
```

Comment:
Please correct me if I am wrong, but based on your description in the README and in your code, this seems to applying a weight `a` to the same nurse, but on sequential days. It looks more like the 
```
2) "hard nurse" constraint: ensures that no nurse works two or more 
   consecutive days.
```
Rather than the "hard shift" constraint.

------
Wouldn't it be more straight forward to use the functions you created to grab the indices?

```
from collections import defaultdict
..
J = defaultdict(int)
for nurse in range(n_nurses):
    for day in range(n_days - 1):
        nurse_day_1 = get_index(nurse, day)
        nurse_day_2 = get_index(nurse, day+1)
        J[nurse_day_1, nurse_day_2] = a
```

This way, the audience doesn't need to try and puzzle out the indices, your helper function does the work for them.

By using default dict, the `J` dictionary will automatically return 0 for any keys that you haven't defined explicitly.
https://docs.python.org/3.3/library/collections.html#collections.defaultdict

Reply:
I love it. This is way more Pythonic. Thanks. I've made this change.

Reply:
Could you also please just confirm that this code is actually the "hard nurse" constraint rather than the "hard shift" constraint? (please see first part of my above comment for context)

Reply:
You were right - thanks for questioning this. I think it is all fixed now.